src/main/certificateManager.js

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+import path from 'path';
+import log from 'electron-log';
+
+import * as WindowManager from './windows/windowManager';
+
+import {addModal} from './modalManager';
+import {getLocalURLString} from './utils';
+
+const modalPreload = path.resolve(__dirname, '../../dist/modalPreload.js');
+
+export class CertificateManager {
+  constructor() {
+    this.certificateQueue = [];
+    this.certificateRequestCallbackMap = new Map();
+  }
+
+  handleSelectCertificate = (event, webContents, url, list, callback) => {
+    if (list.length > 1) {
+      event.preventDefault(); // prevent the app from getting the first certificate available
+
+      // store callback so it can be called with selected certificate
+      this.certificateRequestCallbackMap.set(url, callback);
+      this.addToCertificateQueue(url, list);
+    } else {
+      log.info(`There were ${list.length} candidate certificates. Skipping certificate selection`);
+    }
+  }
+
+  addToCertificateQueue = (url, list) => {
+    this.certificateQueue.push({
+      url,
+      list
+    });
+
+    this.showCertificateModalIfNecessary();
+  }
+
+  showCertificateModalIfNecessary = () => {
+    if (this.certificateQueue.length) {
+      const {url, list} = this.certificateQueue[0];
+      const html = getLocalURLString('certificateModal.html');
+
+      const modalPromise = addModal('certificate', html, modalPreload, {url, list}, WindowManager.getMainWindow());
+      if (modalPromise) {
+        modalPromise.then((data) => {
+          const {cert} = data;
+          this.handleSelectedCertificate(url, cert);
+          this.certificateQueue.shift();
+          this.showCertificateModalIfNecessary();
```

Comment:
Okay, I wasn't sure about the origin of the queues for the authentication methods, but it makes sense that it was just to handle modals. I'll refactor to avoid the queues for auth and certificates.